File: packages/wasmvm/wasmlib/src/state.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use crate::host::*;
use crate::wasmtypes::IKvStore;

pub struct ScImmutableState {}

impl ScImmutableState {
    pub fn exists(&self, key: &[u8]) -> bool {
        state_exists(key)
    }

    pub fn get(&self, key: &[u8]) -> Vec<u8> {
        state_get(key)
    }
}

pub struct ScState {}

impl IKvStore for ScState {
    fn delete(&mut self, key: &[u8]) {
        state_delete(key);
    }

    fn exists(&self, key: &[u8]) -> bool {
        state_exists(key)
    }

    fn get(&self, key: &[u8]) -> Vec<u8> {
        state_get(key)
    }

    fn set(&mut self, key: &[u8], value: &[u8]) {
        state_set(key, value);
    }
}

File: packages/wasmvm/wasmlib/as/wasmlib/state.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {stateDelete, stateExists, stateGet, stateSet} from "./host";
import {IKvStore, Proxy} from "./wasmtypes/proxy";

export class ScImmutableState {
    exists(key: Uint8Array): bool {
        return stateExists(key);
    }

    get(key: Uint8Array): Uint8Array {
        const val = stateGet(key);
        return val === null ? new Uint8Array(0) : val;
    }
}

export class ScState implements IKvStore {
    public static proxy(): Proxy {
        return new Proxy(new ScState());
    }

    delete(key: Uint8Array): void {
        stateDelete(key);
    }

    exists(key: Uint8Array): bool {
        return stateExists(key);
    }

    get(key: Uint8Array): Uint8Array {
        const val = stateGet(key);
        return val === null ? new Uint8Array(0) : val;
    }

    public immutable(): ScImmutableState {
        return this;
    }

    set(key: Uint8Array, value: Uint8Array): void {
        stateSet(key, value);
    }
}
